### PR TITLE
Switch to azblob sdk

### LINKED ISSUE #3698: Add support for backup on Microsoft Azure Blob Storage using Workload Identity
Output of `restic version`
--------------------------
`0.13.0`

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------
In[ Azure AD Workload Identity](https://azure.github.io/azure-workload-identity/docs/) model the Kubernetes cluster becomes a token issuer, issuing tokens to Kubernetes Service Accounts. These service account tokens can be configured to be trusted on Azure AD applications. So Restic shouldn't require the storage access credentials and use the token instead when it doesn't have the credentials. 

What are you trying to do? What problem would this solve?
---------------------------------------------------------
I'm trying to backup some data with restic from a workload deployed in AKS cluster using workload identity federation. But the backup is failing because Restic requires access credentials to store backups on Microsoft Azure Blob Storage. 

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Yes. It's helping us a lot.

### PR DESCRIPTION
**Pull request by ekarlso**
This is the start to fix #3698 since that requires to take a `cred`
struct into many of the client New methods

What does this PR change? What problem does it solve?
-----------------------------------------------------

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

### CODE REVIEW

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
Thanks for the PR. Switching to the newer azure sdk is a good idea. Most comments are smaller issues, the only larger one is the `Save` method.

I'm a bit undecided how the `Save` method should look like. We basically have three options:
- change the case with <256MB files to stage and commit a block similar to saveLarge. The main difference would be that the expected md5 hash of the block is retrieved from rd.Hash()
- Always use `saveLarge`
- Always use `UploadStream`

The simplest way would be to use UploadStream, however, then we'd loose the end-to-end integrity checks for uploaded blocks. My current preference would be to use the first option, even though it's slightly more code.

File: internal/backend/azure/azure.go [hunk 1]
```diff
@@ -41,51 +49,69 @@ var _ restic.Backend = &Backend{}
...
+		// https://github.com/Azure/azure-storage-blob-go/issues/130
 		debug.Log(" - using sas token")
 		sas := cfg.AccountSAS.Unwrap()
-		// strip query sign prefix
-		if sas[0] == '?' {
-			sas = sas[1:]
-		}
-		client, err = storage.NewAccountSASClientFromEndpointToken(url, sas)
+
+		urlWithSAS := fmt.Sprintf("%s/%s?%s", url, cfg.Container, sas)
```
[the remark this answers is not in the document]
Comment (ekarlso) [reply internal/backend/azure/azure.go:85]: That's if you are not acting on a container client directly I think (Though I can certainly test this)

File: internal/backend/azure/azure.go [hunk 2]
```diff
@@ -41,51 +49,69 @@ var _ restic.Backend = &Backend{}
...
 		// Get the client using the SAS Token as authentication, this
 		// is longer winded than above because the SDK wants a URL for the Account
 		// if your using a SAS token, and not just the account name
 		// we (as per the SDK ) assume the default Azure portal.
-		url := fmt.Sprintf("https://%s.blob.core.windows.net/", cfg.AccountName)
+		// https://github.com/Azure/azure-storage-blob-go/issues/130
 		debug.Log(" - using sas token")
 		sas := cfg.AccountSAS.Unwrap()
-		// strip query sign prefix
-		if sas[0] == '?' {
```
[the remark this answers is not in the document]
Comment (ekarlso) [reply internal/backend/azure/azure.go:63]: Yes, point! I added it back

File: internal/backend/azure/azure.go [hunk 3]
```diff
@@ -184,22 +212,24 @@ func (be *Backend) Save(ctx context.Context, h restic.Handle, rd restic.RewindRe
...
 		// CreateBlockBlobFromReader reads length from `Len()``
-		dataReader := azureAdapter{rd}
 
 		// if it's smaller than 256miB, then just create the file directly from the reader
-		ref := be.container.GetBlobReference(objName)
-		ref.Properties.ContentMD5 = base64.StdEncoding.EncodeToString(rd.Hash())
-		err = ref.CreateBlockBlobFromReader(dataReader, nil)
+		bc := be.container.NewBlockBlobClient(objName)
+
+		_, err = bc.UploadStream(ctx, azureAdapter{rd}, &blockblob.UploadStreamOptions{
```
[the remark this answers is not in the document]
Comment (ekarlso) [reply internal/backend/azure/azure.go:225]: > Thanks for the PR. Switching to the newer azure sdk is a good idea. Most comments are smaller issues, the only larger one is the `Save` method.
> 
> I'm a bit undecided how the `Save` method should look like. We basically have three options:
> 
>     * change the case with <256MB files to stage and commit a block similar to saveLarge. The main difference would be that the expected md5 hash of the block is retrieved from rd.Hash()
> 
>     * Always use `saveLarge`
> 
>     * Always use `UploadStream`
> 
> 
> The simplest way would be to use UploadStream, however, then we'd loose the end-to-end integrity checks for uploaded blocks. My current preference would be to use the first option, even though it's slightly more code.

I went the route you said and duplicated the saveLarge behaviour

File: internal/backend/azure/azure.go [hunk 4]
```diff
@@ -41,51 +49,69 @@ var _ restic.Backend = &Backend{}
...
+
+		urlWithSAS := fmt.Sprintf("%s/%s?%s", url, cfg.Container, sas)
+		client, err = azContainer.NewClientWithNoCredential(urlWithSAS, opts)
 		if err != nil {
 			return nil, errors.Wrap(err, "NewAccountSASClientFromEndpointToken")
 		}
 	} else {
-		return nil, errors.New("no azure authentication information found")
-	}
+		cred, err := azidentity.NewDefaultAzureCredential(nil)
```
[the remark this answers is not in the document]
Comment (ekarlso) [reply internal/backend/azure/azure.go:91]: Removed from this PR

File: internal/backend/azure/azure.go [hunk 5]
```diff
@@ -241,26 +266,23 @@ func (be *Backend) saveLarge(ctx context.Context, objName string, rd restic.Rewi
 		// upload it as a new "block", use the base64 hash for the ID
 		h := md5.Sum(buf)
 		id := base64.StdEncoding.EncodeToString(h[:])
-		debug.Log("PutBlock %v with %d bytes", id, len(buf))
-		err = file.PutBlock(id, buf, &storage.PutBlockOptions{ContentMD5: id})
+
+		reader := bytes.NewReader(buf)
+		debug.Log("StageBlock %v with %d bytes", id, len(buf))
+		_, err = blockBlobClient.StageBlock(ctx, id, streaming.NopCloser(reader), &blockblob.StageBlockOptions{})
```
[the remark this answers is not in the document]
Comment (ekarlso) [reply internal/backend/azure/azure.go:272]: Thanks, added.

File: internal/backend/azure/azure.go [hunk 6]
```diff
@@ -184,22 +212,24 @@ func (be *Backend) Save(ctx context.Context, h restic.Handle, rd restic.RewindRe
...
 
 		// if it's smaller than 256miB, then just create the file directly from the reader
-		ref := be.container.GetBlobReference(objName)
-		ref.Properties.ContentMD5 = base64.StdEncoding.EncodeToString(rd.Hash())
-		err = ref.CreateBlockBlobFromReader(dataReader, nil)
+		bc := be.container.NewBlockBlobClient(objName)
+
+		_, err = bc.UploadStream(ctx, azureAdapter{rd}, &blockblob.UploadStreamOptions{
+			HTTPHeaders: &blob.HTTPHeaders{
+				BlobContentMD5: []byte(rd.Hash()),
```
[the remark this answers is not in the document]
Comment (ekarlso) [reply internal/backend/azure/azure.go:227]: Removed UploadStream usage all together and duplicated code in saveLarge as you appointed.

File: internal/backend/azure/azure.go [hunk 7]
```diff
@@ -285,46 +307,55 @@ func (be *Backend) openReader(ctx context.Context, h restic.Handle, length int,
...
 	be.sem.GetToken()
-	err := blob.GetProperties(nil)
-	be.sem.ReleaseToken()
+	props, err := blob.GetProperties(ctx, nil)
 
 	if err != nil {
-		debug.Log("blob.GetProperties err %v", err)
-		return restic.FileInfo{}, errors.Wrap(err, "blob.GetProperties")
+		if bloberror.HasCode(err, bloberror.BlobNotFound) {
+			return restic.FileInfo{}, nil
```
[the remark this answers is not in the document]
Comment (ekarlso) [reply internal/backend/azure/azure.go:349]: Removed the specifc BlobNotFound case

File: internal/backend/azure/azure.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ekarlso) [reply internal/backend/azure/azure.go]: I fixed this and tested using SAS and it works fine.

**Comment by MichaelEischer** [comment]
I've fixed the remaining review comments. Apparently the azure SDK requires golang 1.18, which would be major bump in the required golang version...

**Comment by ekarlso** [comment]
@MichaelEischer idk what's up here but I get 
```
debug log file test.log
debug enabled
Fatal: create key in repository at azure:test:/ failed: StageBlock: PUT https://xxxblob.core.windows.net/test/keys/5a09ec71a89cf36911982a9de23ce53639436830a4114662fe2c457937917c9c
--------------------------------------------------------------------------------
RESPONSE 400: 400 The MD5 value specified in the request did not match with the MD5 value calculated by the server.
ERROR CODE: Md5Mismatch
--------------------------------------------------------------------------------
<?xml version="1.0" encoding="utf-8"?><Error><Code>Md5Mismatch</Code><Message>The MD5 value specified in the request did not match with the MD5 value calculated by the server.
RequestId:afd69da6-501e-0038-77cf-01c51e000000
Time:2022-11-26T19:43:01.9531956Z</Message><UserSpecifiedMd5>MBrCsbYtOgytzT9iexEK4w==</UserSpecifiedMd5><ServerCalculatedMd5>H1A55QvWaykMVmhNhVDGwg==</ServerCalculatedMd5></Error>
```

Seems we're calculating or giving a mismach of the MD5?

File: internal/backend/azure/azure.go [hunk 8]
```diff
@@ -184,22 +196,38 @@ func (be *Backend) Save(ctx context.Context, h restic.Handle, rd restic.RewindRe
...
 		// if it's smaller than 256miB, then just create the file directly from the reader
-		ref := be.container.GetBlobReference(objName)
-		ref.Properties.ContentMD5 = base64.StdEncoding.EncodeToString(rd.Hash())
-		err = ref.CreateBlockBlobFromReader(dataReader, nil)
+		blockBlobClient := be.container.NewBlockBlobClient(objName)
+
+		// upload it as a new "block", use the base64 hash for the ID
+		id := base64.StdEncoding.EncodeToString(rd.Hash())
+
+		buf := make([]byte, saveLargeSize)
```
Comment (MichaelEischer) [inline internal/backend/azure/azure.go:212]: The buffer is never filled with data. It should probably be:

```suggestion
		buf := make([]byte, rd.Length())
		_, err = io.ReadFull(rd, buf)
		if err != nil {
			return errors.Wrap(err, "ReadFull")
		}
```

**Comment by ekarlso** [comment]
So now it's working that far but now it hangs on:
```processed 1 files, 8.232 KiB in 0:00
snapshot 709fadfc saved
``` 

I guess it's due to one of the locks not releasing?

**Comment by MichaelEischer** [comment]
> So now it's working that far but now it hangs on:
> 
> ```
> snapshot 709fadfc saved
> ```
> 
> I guess it's due to one of the locks not releasing?

Without more context that's usually pretty hard to tell. In case of a hanging golang program it's often helpful to send a SIGQUIT to the program (ctrl+\). Then the runtime prints a stacktrace of all goroutines and exits. Here it was probably caused by a leaked token in the `Save` method.

**Comment by ekarlso** [comment]
No but I well def use it before pushing more code! Thanks for the tip

**Comment by MichaelEischer** [comment]
> goroutine 43 [chan send]:
> github.com/restic/restic/internal/backend/sema.Semaphore.GetToken(...)
>         /home/ekarlso/projects/evry/dds/infrastructure/docker-images/restic/internal/backend/sema/semaphore.go:27
> github.com/restic/restic/internal/backend/azure.(*Backend).List(0xc00011d290, {0xc39138, 0xc00019e140}, 0x1, 0xc0004f4080)
>         /home/ekarlso/projects/evry/dds/infrastructure/docker-images/restic/internal/backend/azure/azure.go:426 +0x35e

I have no idea which version of the code that is. But it looks like it does not include the last three bugfix commits which I've added to this PR (otherwise the line number doesn't make sense).

**Comment by ekarlso** [comment]
@MichaelEischer all tests are green and the branch has been updated. Thanks for your help!

**Review by MichaelEischer — APPROVED** [review]
LGTM. I've squashed the commits (the bugfix commits probably aren't useful anyways) and removed the changelog.

**Pull request merged**